module/signature/random_beacon_signer_store.go

```diff
@@ -0,0 +1,109 @@
+package signature
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/storage"
+)
+
+type EpochAwareRandomBeaconSignerStore struct {
+	epochLookup module.EpochLookup          // used to fetch epoch counter by view
+	keys        storage.DKGKeys             // used to fetch DKG private key by epoch
+	signers     map[uint64]module.MsgSigner // cache of signers by epoch
+}
+
+func NewEpochAwareRandomBeaconSignerStore(epochLookup module.EpochLookup, keys storage.DKGKeys) *EpochAwareRandomBeaconSignerStore {
+	return &EpochAwareRandomBeaconSignerStore{
+		epochLookup: epochLookup,
+		keys:        keys,
+		signers:     make(map[uint64]module.MsgSigner),
+	}
+}
+
+// GetSigner returns the threshold-signer for signing objects at a
+// given view. The view determines the epoch, which determines the DKG private
+// key underlying the signer.
+// It returns:
+//  - (signer, nil) if DKG was completed in the epoch of the view, signer is not nil
+//  - (nil, protocol.ErrEpochNotCommitted) if no epoch found for given view
+//  - (nil, DKGIncompleteError) if DKG was not completed in the epoch of the view
+//  - (nil, error) if there is any exception
+func (s *EpochAwareRandomBeaconSignerStore) GetSigner(view uint64) (module.MsgSigner, error) {
+	// fetching the epoch by view, if epoch is found, then DKG must have been completed
+	epoch, err := s.epochLookup.EpochForViewWithFallback(view)
```

Comment:
I am worried about the performance here. To the best of my understanding, this _always_ hits the data base which will impact the hot path of consensus. 

We can utilize some knowledge about the signing process to optimize the implementation:
* a consensus replica only signs (at most) once for its current view. Hence, `view` increases strictly monotonously.  
* Chances are overwhelmingly high that I am _exactly_ in the same epoch as during the last call `GetSigner`, as Epoch changes are relatively rare. 

Therefore, it is totally sufficient to cache 
* the _signer_ from the last call, along with some epoch details:
   * first view of epoch and last view of epoch
   * thereby, I can easily determine whether I am still in the same epoch
   * if I get a view _before_ the current epoch, I can just error (unexpected critical)
   * If I am in a view _beyond_ the view of the current epoch, I load a new signer and cache it.